### PR TITLE
Adding a custom http handler for read/write access validation

### PR DESCRIPTION
**Pull request by jjbustamante**
Signed-off-by: Juan Bustamante <jbustamante@vmware.com>

## Summary

This PR is adding a custom HTTP handler to the docker registry to validate read/write access to some image name, these methods will be used to fix the issue [572 - validate registry credentials](https://github.com/buildpacks/lifecycle/issues/572) in the [lifecycle](https://github.com/buildpacks/lifecycle) repository

## Documentation

- Should this change be documented?
    - [x] No

## Related
Resolves #572 (in lifecycle)

### CODE REVIEW

File: testhelpers/docker_registry.go
Comment (natalieparellano) [inline testhelpers/docker_registry.go:30]: Do we need this? Is it set anywhere?

File: testhelpers/docker_registry.go
Comment (natalieparellano) [inline testhelpers/docker_registry.go:152]: Do you mean to be setting `r.authnHandler` here?

Comment (natalieparellano) [reply testhelpers/docker_registry.go:152]: (and if so, should we break after this line)

File: testhelpers/docker_registry.go
Comment (natalieparellano) [inline testhelpers/docker_registry.go:113]: A helper function might be useful here (to make the code more readable)

**Comment by natalieparellano** [comment]
I love it! Left a couple nits & questions.

File: testhelpers/docker_registry.go
Comment (jjbustamante) [reply testhelpers/docker_registry.go:152]: No,  actually I am wrapping the regular handler with my custom behavior like this:
wrapping
The idea is to keep the basic authentication logic for the registry, then in the cases, I want to simulate the read/write behavior access I will do it in the custom and all other requests I do not care about will go to the normal registry logic. So, lines 151 and 152 are ok exactly like that.

File: testhelpers/docker_registry.go
Comment (jjbustamante) [reply testhelpers/docker_registry.go:30]: Oh. I just noticed I am not using it! let me removes it

File: testhelpers/docker_registry.go
Comment (aemengo) [reply testhelpers/docker_registry.go:113]: I'm having trouble here. `name` is supposed to be the image tag, correct? The same value passed into the `permissions` mapping?

The following is what I'm seeing for `r.URL.Path` with `remote.NewImage()`. Am I using it right?

```shell
# image tag: 192.168.1.164:63737/aemengo/test
Path: /v2/aemengo/test/blobs/sha256:60df352a35c779a36da39e670ddc90742f64209045365ea0ac4f5eb83422a615
Path: /v2/aemengo/test/blobs/uploads/
Path: /v2/aemengo/test/blobs/uploads/4831389563158288344
Path: /v2/aemengo/test/blobs/uploads/4831389563158288344
Path: /v2/aemengo/test/manifests/latest
Path: /v2/aemengo/test/manifests/latest
Path: /v2/aemengo/test/blobs/uploads/
Path: /v2/aemengo/test/blobs/uploads/261049867304784443
```

Comment (jjbustamante) [reply testhelpers/docker_registry.go:113]: well, in this case, `name` should be `test` and that should be the key to find in `permissions`. that you put before it is not going to be used to determine the permission (at least from the test perspective)

File: testhelpers/docker_registry.go
Comment (natalieparellano) [reply testhelpers/docker_registry.go:152]: Ohhh thank you for this explanation! That makes sense. Maybe add a comment?

Comment (natalieparellano) [reply testhelpers/docker_registry.go:152]:
```suggestion
			// wrap registry handler with custom handler
			r.regHandler = Custom(r.regHandler, r.customPrivileges)
```

File: testhelpers/docker_registry.go
Comment (natalieparellano) [reply testhelpers/docker_registry.go:113]: Hmm... this playground is giving me a blank name: https://play.golang.org/p/Zn2EqfiHGb5

I think we can't assume that `test` will be at the end of the path...

Comment (jjbustamante) [reply testhelpers/docker_registry.go:113]: Yeap, it's my mistake I will fix it ;) thanks @natalieparellano and @aemengo

Comment (jjbustamante) [reply testhelpers/docker_registry.go:113]: @aemengo , @natalieparellano 

What about something like [this](https://play.golang.org/p/QfbivVx2xIt) to parse the URL and extract the `name`, @aemengo I am trying to handle you mentioned 😄 .

I like the regex to isolate the string to V2/[whatever]/(blobs|manifest|tag), but I open to suggestion to get [whatever] in a optimal way

Comment (aemengo) [reply testhelpers/docker_registry.go:113]: @jjbustamante Why so complicated? 😅 Take a look at [this one](https://play.golang.org/p/F3b-dSljAQT)

Comment (jjbustamante) [reply testhelpers/docker_registry.go:113]: hahaha.. come on! I hate regex, thanks! any thoughts on best option to remove the /V2/****/(blobs|manifest|tag) stuff? and just keep the name?

Comment (aemengo) [reply testhelpers/docker_registry.go:113]: If you use the `FindStringSubmatch` method, the name will be 2nd value in an array of 2 elements. 👍🏾 

```
matches := s.FindStringSubmatch(str)
matches[1] //<--- name
```

Comment (jjbustamante) [reply testhelpers/docker_registry.go:113]: Awesome! I like it! thanks

File: testhelpers/docker_registry.go
Comment (jjbustamante) [reply testhelpers/docker_registry.go:137]: Added at the following [commit](https://github.com/buildpacks/imgutil/pull/125/commits/4261e2896afc176ea6a4967a4bd255f780340892)

File: testhelpers/docker_registry.go
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Removed in the following [commit](https://github.com/buildpacks/imgutil/pull/125/commits/4261e2896afc176ea6a4967a4bd255f780340892)

File: testhelpers/docker_registry.go
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: I fixed the issue and took into consideration the suggestion, the logic was moved to the helper function and I used the regex pattern matching to find the `name` see [commit](https://github.com/buildpacks/imgutil/pull/125/commits/4261e2896afc176ea6a4967a4bd255f780340892)

File: testhelpers/image_privileges.go
Comment (aemengo) [inline testhelpers/image_privileges.go:24]:
```suggestion
func NewImagePrivileges(imageName string) (priv ImagePrivileges) {
	priv.readable = strings.Contains(imageName, "readable")
	priv.writeable = strings.Contains(imageName, "writable")
	return
}
```

I liked the old interface better. But you're using it, not me 😅

**Review by aemengo — APPROVED** [review]

File: testhelpers/image_privileges.go
Comment (yaelharel) [reply testhelpers/image_privileges.go:24]: I don't see who is using this function but I think the `if` conditions here are redundant.
I'm ok with returning a pointer, but I would remove the conditions and just use the return value from `strings.Contains` to set the variables.

Comment (jjbustamante) [reply testhelpers/image_privileges.go:24]: I added this way of setting the properties based on @natalieparellano comments [here](https://github.com/buildpacks/lifecycle/pull/646#discussion_r669816609), I will use this method in the lifecycle unit tests

Comment (jjbustamante) [reply testhelpers/image_privileges.go]: Fixed in the following [commit](https://github.com/buildpacks/imgutil/pull/125/commits/b6d4eb2c54b191a59bf4a46c43c492a881e6889b)

File: testhelpers/docker_registry.go
Comment (yaelharel) [inline testhelpers/docker_registry.go]: This is a map from an `imageName` to its permissions, right?
I think it's worth adding it as a comment

File: testhelpers/image_privileges.go
Comment (yaelharel) [inline testhelpers/image_privileges.go:21]: Can you please add a description to this function?

File: testhelpers/docker_registry.go
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Added in the following [commit](https://github.com/buildpacks/imgutil/pull/125/commits/714d0d197edf3ff7391210592d4c3552cdfca10b)

File: testhelpers/image_privileges.go
Comment (jjbustamante) [reply testhelpers/image_privileges.go:26]: Added in the following [commit](https://github.com/buildpacks/imgutil/pull/125/commits/714d0d197edf3ff7391210592d4c3552cdfca10b)

File: testhelpers/docker_registry.go
Comment (yaelharel) [inline testhelpers/docker_registry.go:105]: Looking at the return value of `ReadOnly` (although the use case is a bit different), I'm wondering whether it should be 401 or 405 ("Method Not Allowed")? I'm not sure which one is more accurate in this case.

File: testhelpers/docker_registry.go
Comment (yaelharel) [inline testhelpers/docker_registry.go:103]: I would add a comment in each of these conditions to make the flow clearer. Something like:
```
// read request was arrived while read permission isn't allowed
```

**Review by yaelharel — COMMENTED** [review]
I think it's worth adding some tests of this new functionality.

File: testhelpers/docker_registry.go
Comment (yaelharel) [inline testhelpers/docker_registry.go:29]:
```suggestion
	customPrivileges map[string]ImagePrivileges // map from an imageName to its permissions
```

File: testhelpers/docker_registry.go
Comment (jjbustamante) [reply testhelpers/docker_registry.go:105]: Well, that's actually one of the problems with the ReadOnly when we try to validate read/write access, we are expecting to receive `401 Unauthorized` if your credentials are wrong or you don't have access to resources.

File: testhelpers/docker_registry.go
Comment (jjbustamante) [reply testhelpers/docker_registry.go:103]: Done. [commit](https://github.com/buildpacks/imgutil/pull/125/commits/a0fd89a42b3519eaf6dc613931da5b34d11cf397)

File: testhelpers/docker_registry.go
Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Done [commit](https://github.com/buildpacks/imgutil/pull/125/commits/a0fd89a42b3519eaf6dc613931da5b34d11cf397)

File: testhelpers/docker_registry.go
Comment (yaelharel) [reply testhelpers/docker_registry.go:106]: So why don't we fix the `ReadOnly` function?

File: testhelpers/docker_registry.go
Comment (yaelharel) [inline testhelpers/docker_registry.go:111]:
```suggestion
					// write request was arrived while write permission isn't allowed
```

Comment (jjbustamante) [reply testhelpers/docker_registry.go]: Fixed [commit](https://github.com/buildpacks/imgutil/pull/125/commits/6a95fbcf1491910ebe83969ffeec0d73807786de) 

Sorry!

**Comment by jjbustamante** [comment]
> I think it's worth adding some tests of this new functionality.

I just added more unit tests to the read/write methods using the custom handler. I noticed that I think the `checkReadWriteAccess` was not validating read access, I will like to take a look at that test case (not sure if that is even possible, having write access but not read, I think @natalieparellano mentioned that before). Any thoughts will be awesome.

File: testhelpers/docker_registry.go
Comment (jjbustamante) [reply testhelpers/docker_registry.go:106]: There are several tests that depend on that, it could be an option, but actually not sure how good or bad it could be, but in any case, this docker registry implementation is a custom implementation created just to handle test scenarios. From my perspective having a version of a registry that behaves in a certain way for given scenarios is not that bad

Comment (natalieparellano) [reply testhelpers/docker_registry.go:106]: IIUC we expect Docker Hub to return a 401, therefore the test registry should return a 401.

File: remote/remote.go
Comment (natalieparellano) [inline remote/remote.go:359]: I agree that "write only" access would be silly. But this doesn't do any harm...

File: testhelpers/image_privileges.go
Comment (natalieparellano) [inline testhelpers/image_privileges.go:15]: Could we add a docs comment here? Keying the permissions off the image name (while convenient) wasn't immediately obvious to me.

File: remote/remote_test.go
Comment (natalieparellano) [inline remote/remote_test.go:1683]: Should we check read access here?

File: remote/remote_test.go
Comment (natalieparellano) [inline remote/remote_test.go:1699]: Should we check read access here?

**Comment by natalieparellano** [comment]
@jjbustamante thank you for adding the tests! This looks great!

**Review by natalieparellano — APPROVED** [review]

File: remote/remote_test.go
Comment (jjbustamante) [reply remote/remote_test.go:1683]: Do you mean adding the assert `image.CheckReadAccess()`? because that one seems to be tested at line 1585

File: remote/remote_test.go
Comment (jjbustamante) [reply remote/remote_test.go:1699]: Thinking a lit bit more about it, with these new docker registry handlers, I can actually cut in half the code and added the assertions for `CheckReadWriteAccess()` and `CheckReadAccess()` in the same test scenarios. Would it help on making the test cases more readable?

File: remote/remote.go
Comment (jjbustamante) [reply remote/remote.go:359]: Actually, because we changed the method signature to `CheckReadWriteAccess()`, actually it should also validate the Read access, right?

File: remote/remote_test.go
Comment (natalieparellano) [reply remote/remote_test.go:1683]: Ohh I missed that these were all scoped to `CheckReadAccess()`. I think it's okay.

File: remote/remote_test.go
Comment (natalieparellano) [reply remote/remote_test.go:1699]: Oh, maybe! Up to you...

File: testhelpers/docker_registry.go
Comment (yaelharel) [reply testhelpers/docker_registry.go:106]: @natalieparellano, it makes sense that it'll return 401. My question is why `ReadOnly` returns 405 🤔

File: testhelpers/image_privileges.go
Comment (jjbustamante) [reply testhelpers/image_privileges.go:21]: Added in this [commit](https://github.com/buildpacks/imgutil/pull/125/commits/2486b56700226982aa19e6e383bba488b80e0394)

File: testhelpers/image_privileges.go
Comment (yaelharel) [inline testhelpers/image_privileges.go:25]: Maybe I'm wrong but according to what I saw in our code, we usually don't set the returned variable name in the function definition. We define it as part of the function and return it explicitly.
If I'm right, maybe it's better to change the function to:

```suggestion
func NewImagePrivileges(imageName string) (priv ImagePrivileges) {
	image ImagePrivileges
	image.readable = strings.Contains(imageName, "readable")
	image.writable = strings.Contains(imageName, "writable")
	return image
}
```

File: remote/remote_test.go
Comment (yaelharel) [inline remote/remote_test.go:27]:
```suggestion
	onlyReadImage    = "image-readable"
```

File: remote/remote_test.go
Comment (yaelharel) [inline remote/remote_test.go:28]:
```suggestion
	onlyWriteImage    = "image-writable"
```

File: remote/remote_test.go
Comment (yaelharel) [inline remote/remote_test.go:29]:
```suggestion
	noAccessImage = "no-access-image"
```

File: remote/remote_test.go
Comment (yaelharel) [inline remote/remote_test.go:61]: Maybe you can define all of the consts in a map, and then loop on that map and initialize the `customPrivileges`.
In this way, if we'll decide to add/remove images, we'll need to do it only in one place.

File: remote/remote_test.go
Comment (yaelharel) [inline remote/remote_test.go:1573]:
```suggestion
			when("image has read and write access", func() {
```

Comment (yaelharel) [reply remote/remote_test.go:1573]: Why do we only call `CheckReadAccess` instead of `CheckReadWriteAccess`?
Moreover, maybe it's better that `CheckReadWriteAccess` will return 2 boolean values for the read and write access separately.

File: remote/remote_test.go
Comment (yaelharel) [inline remote/remote_test.go:1671]:
```suggestion
			when("image has read and write access", func() {
```

File: remote/remote_test.go
Comment (jjbustamante) [reply remote/remote_test.go]: Added in the following [commit](https://github.com/buildpacks/imgutil/pull/125/commits/87d5ba9fc355d30923e2cf672f08e02c9de908fd)

File: remote/remote_test.go
Comment (jjbustamante) [reply remote/remote_test.go]: Added in the following [commit](https://github.com/buildpacks/imgutil/pull/125/commits/87d5ba9fc355d30923e2cf672f08e02c9de908fd)

File: remote/remote_test.go
Comment (jjbustamante) [reply remote/remote_test.go]: Added in the following [commit](https://github.com/buildpacks/imgutil/pull/125/commits/87d5ba9fc355d30923e2cf672f08e02c9de908fd)

File: remote/remote_test.go
Comment (jjbustamante) [reply remote/remote_test.go]: Added in the following [commit](https://github.com/buildpacks/imgutil/pull/125/commits/87d5ba9fc355d30923e2cf672f08e02c9de908fd)

File: remote/remote_test.go
Comment (jjbustamante) [reply remote/remote_test.go]: Added in the following [commit](https://github.com/buildpacks/imgutil/pull/125/commits/87d5ba9fc355d30923e2cf672f08e02c9de908fd)

File: remote/remote.go
Comment (yaelharel) [reply remote/remote.go:359]: I have some thoughts about the `CheckReadWriteAccess` and `CheckReadAccess` functions:
* Maybe we should move `remote.CheckPushPermission(ref, i.keychain, http.DefaultTransport) == nil` to a private function `checkWriteAccess` so the code will be clearer.
* Should `CheckReadWriteAccess` return two booleans or a struct with two booleans so it'll be clearer what kind of access (if any) the image has?
* Although both functions are public, I think it's redundant to run almost the same tests as part of `remote_test.go`. Should we only keep those under `CheckReadWriteAccess`? And if not, ,maybe we can create a function to test all of these scenarios together (the function will get the function that should run on the image - either `CHeckReadWriteAccess` or `CheckReadAccess`).

Comment (jjbustamante) [reply remote/remote.go:359]: Regarding your third comment, I thought something similar [here](https://github.com/buildpacks/imgutil/pull/125#discussion_r671434624). This is a preview of how it will look if I move the `CheckReadAccess` assertions to the `CheckReadWriteAccess` scenarios

```
when("image has read and write access", func() {
    it("returns true", func() {
	    image, err := remote.NewImage(customRegistry.RepoName(readWriteImage), authn.DefaultKeychain)
	    h.AssertNil(t, err)
	    h.AssertEq(t, image.CheckReadWriteAccess(), true)
	    h.AssertEq(t, image.CheckReadAccess(), true)
    })
})

when("image has read access but no write access", func() {
    it("returns true for CheckReadAccess() and false for CheckReadWriteAccess()", func() {
	    image, err := remote.NewImage(customRegistry.RepoName(onlyReadImage), authn.DefaultKeychain)
	    h.AssertNil(t, err)
	    h.AssertEq(t, image.CheckReadWriteAccess(), false)
	    h.AssertEq(t, image.CheckReadAccess(), true)
    })
})

when("image doesn't have read access but has write access", func() {
    it("returns false", func() {
	    image, err := remote.NewImage(customRegistry.RepoName(onlyWriteImage), authn.DefaultKeychain)
	    h.AssertNil(t, err)
	    h.AssertEq(t, image.CheckReadWriteAccess(), false)
	    h.AssertEq(t, image.CheckReadAccess(), false)
    })
})

when("image doesn't have read nor write access", func() {
    it("returns false", func() {
	    image, err := remote.NewImage(customRegistry.RepoName(noAccessImage), authn.DefaultKeychain)
	    h.AssertNil(t, err)
	    h.AssertEq(t, image.CheckReadWriteAccess(), false)
	    h.AssertEq(t, image.CheckReadAccess(), false)
    })
})
```

Comment (yaelharel) [reply remote/remote.go:359]: I'm wondering if it doesn't make the code confusing, because all of these `when`s will be under `when("CheckReadWriteAccess"`, right?
I thought about completely deleting the `CheckReadAccess` calls because they're part of the `CheckReadWriteAccess` code, but then I think that we need to change a bit the `CheckReadWriteAccess` interface to return what kind of access we have.

File: testhelpers/docker_registry.go
Comment (natalieparellano) [reply testhelpers/docker_registry.go:106]: I am not sure - IIRC @micahyoung had done some investigation on what the failure response looks like for different registries. The [distribution spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) is a bit ambiguous on what specific code should be returned. I would not be opposed to updating `ReadOnly`, but I don't think it needs to block this PR.

Comment (yaelharel) [reply testhelpers/docker_registry.go:106]: Of course.
Maybe it's worth opening a separate issue for it - fix the errors that should be returned in each case (so at least we'll be consistent).

**Pull request merged**
